File: tests/agents/memory/TestDataBuffer.py
import pytest
import math

import torch

from benchmarks.agents.memory.FastDataBuffer import FastDataBuffer as DataBuffer
from benchmarks.agents.memory.ReplayBuffer import Experience


class TestDataBuffer:
    """
    Class testing the data buffer.
    """

    @staticmethod
    def n_steps_reward(r_0, increment, gamma, n):
        r = 0
        for i in range(n):
            r += math.pow(gamma, i) * (r_0 + i * increment)
        return r

    @pytest.mark.parametrize("capacity, n_steps, gamma", [
        (5, 1, 1), (5, 1, 0.9), (5, 2, 1), (5, 2, 0.99),
        (6, 2, 0.95), (7, 1, 0.5), (8, 3, 0.75), (9, 2, 0.8),
        (5, 6, 0.979), (5, 1, 0.98), (5, 1, 0.999), (9, 8, 0.1),
    ])
    def test_storing_and_retrieval_multiple_episodes(self, capacity, n_steps, gamma):

        # Create a data buffer.
        buffer = DataBuffer(capacity=capacity, n_steps=n_steps, gamma=gamma, initial_priority=1, n_children=10)

        # Create the experiences at time t.
        experiences = [
            Experience(obs=torch.zeros([4, 10, 10]), action=t, reward=t, done=(t == capacity - 1), next_obs=torch.zeros([4, 10, 10]))
            for t in range(capacity)
        ] + [
            Experience(obs=torch.zeros([4, 10, 10]), action=t, reward=t, done=(t == capacity - 1), next_obs=torch.zeros([4, 10, 10]))
            for t in range(capacity - 1)
        ]

        # Create the multistep experiences at time t (experiences expected to be returned by the frame buffer).
        result_experiences = []
        for _ in range(2):
            for t in range(capacity):
                tn = min(t + n_steps, capacity)
                tn1 = min(t + n_steps - 1, capacity - 1)
                result_experiences.append(Experience(
                    obs=torch.zeros([4, 10, 10]), action=t, reward=self.n_steps_reward(t, 1, gamma, tn - t),
                    done=(tn1 == capacity - 1), next_obs=torch.zeros([4, 10, 10])
                ))

        # Fill the buffer with experiences.
        for t in range(capacity):
            buffer.append(experiences[t])

        # Check that experiences in the frame buffer are as expected.
        indices = torch.tensor([t for t in range(capacity)])
        actions, rewards, dones = buffer[indices]
        for t in range(capacity):
            assert actions[t] == result_experiences[t].action
            assert abs(rewards[t].item() - result_experiences[t].reward) < 1e-5
            assert dones[t] == result_experiences[t].done

        # Keep pushing experiences to the buffer, effectively replacing all experiences in the frame buffer.
        for t in range(capacity - 1):
            buffer.append(experiences[capacity + t])

        # Check that the new experiences in the frame buffer are as expected.
        indices = torch.tensor([t for t in range(capacity)])
        actions, rewards, dones = buffer[indices]
        for t in range(capacity - n_steps + 1):
            assert actions[t] == result_experiences[capacity - n_steps + t].action
            assert abs(rewards[t].item() - result_experiences[capacity - n_steps + t].reward) < 1e-5
            assert dones[t] == result_experiences[capacity - n_steps + t].done

    @pytest.mark.parametrize("capacity, n_steps, gamma", [
        (5, 1, 1), (5, 1, 0.9), (5, 2, 1), (5, 2, 0.99),
        (6, 2, 0.95), (7, 1, 0.5), (8, 3, 0.75), (9, 2, 0.8),
        (5, 6, 0.979), (5, 1, 0.98), (5, 1, 0.999), (9, 8, 0.1),
    ])
    def test_storing_and_retrieval(self, capacity, n_steps, gamma):

        # Create a data buffer.
        buffer = DataBuffer(capacity=capacity, n_steps=n_steps, gamma=gamma, initial_priority=1, n_children=10)

        # Create the experiences at time t.
        experiences = [
            Experience(obs=torch.zeros([4, 10, 10]), action=t, reward=t, done=False, next_obs=torch.zeros([4, 10, 10]))
            for t in range(2 * capacity + n_steps - 1)
        ]

        # Create the multistep experiences at time t (experiences expected to be returned by the data buffer).
        result_experiences = [
            Experience(obs=torch.zeros([4, 10, 10]), action=t, reward=self.n_steps_reward(t, 1, gamma, n_steps), done=False, next_obs=torch.zeros([4, 10, 10]))
            for t in range(2 * capacity)
        ]

        # Fill the buffer with experiences.
        n_experiences = capacity + n_steps - 1
        for t in range(n_experiences):
            buffer.append(experiences[t])

        # Check that experiences in the frame buffer are as expected.
        indices = torch.tensor([t for t in range(capacity)])
        actions, rewards, dones = buffer[indices]
        for t in range(capacity):
            assert actions[t] == result_experiences[t].action
            assert abs(rewards[t].item() - result_experiences[t].reward) < 1e-5
            assert dones[t] == result_experiences[t].done

        # Keep pushing experiences to the buffer, effectively replacing all experiences in the frame buffer.
        for t in range(capacity):
            buffer.append(experiences[t + n_experiences])

        # Check that the new experiences in the frame buffer are as expected.
        indices = torch.tensor([t for t in range(capacity)])
        actions, rewards, dones = buffer[indices]
        for t in range(capacity):
            assert actions[t] == result_experiences[t + capacity].action
            assert abs(rewards[t].item() - result_experiences[t + capacity].reward) < 1e-5
            assert dones[t] == result_experiences[t + capacity].done

File: tests/src/agents/memory/test_data_buffer.cpp
// Copyright 2025 Theophile Champion. No Rights Reserved.

#include "agents/memory/test_data_buffer.hpp"
#include "agents/memory/data_buffer.hpp"
#include "relab_test.hpp"
#include <gtest/gtest.h>
#include <memory>
#include <torch/extension.h>

using namespace relab::agents::memory;

namespace relab::test::agents::memory {

DataBufferParameters::DataBufferParameters(int capacity, int n_steps,
                                           float gamma, float initial_priority,
                                           int n_children)
    : capacity(capacity), n_steps(n_steps), gamma(gamma),
      initial_priority(initial_priority), n_children(n_children) {}

DataBufferParameters::DataBufferParameters()
    : DataBufferParameters(0, 0, 0, 0, 0) {}

void TestDataBuffer::SetUp() {

  // Create the replay buffer.
  this->params = GetParam();
  this->buffer = std::make_unique<DataBuffer>(
      this->params.capacity, this->params.n_steps, this->params.gamma,
      this->params.initial_priority, this->params.n_children);

  // Create the observations at time t.
  int n_observations = 2 * this->params.capacity + this->params.n_steps;
  this->observations = getObservations(n_observations, 1, 4);
}

TEST_P(TestDataBuffer, TestStoringAndRetrievalMultipleEpisodes) {

  // Create the experiences at time t.
  auto experiences =
      getExperiences(observations, 2 * params.capacity - 1, params.capacity);

  // Create the multistep experiences at time t (experiences expected to be
  // returned by the frame buffer).
  auto results =
      getResultExperiences(observations, params.gamma, params.n_steps,
                           2 * params.capacity, params.capacity);

  // Fill the buffer with experiences.
  for (int t = 0; t < params.capacity; t++) {
    buffer->append(experiences[t]);
  }

  // Check that experiences in the frame buffer are as expected.
  auto indices = torch::arange(params.capacity);
  auto [actions, rewards, dones] = (*buffer)[indices];
  for (int t = 0; t < params.capacity; t++) {
    EXPECT_EQ(actions[t].item<int>(), results[t].action);
    EXPECT_TRUE(std::abs(rewards[t].item<float>() - results[t].reward) <
                TEST_EPSILON);
    EXPECT_EQ(dones[t].item<bool>(), results[t].done);
  }

  // Keep pushing experiences to the buffer, effectively replacing all
  // experiences in the frame buffer.
  for (int t = 0; t < params.capacity - 1; t++) {
    buffer->append(experiences[params.capacity + t]);
  }

  // Check that the new experiences in the frame buffer are as expected.
  std::tie(actions, rewards, dones) = (*buffer)[indices];
  for (int t = 0; t < params.capacity - params.n_steps + 1; t++) {
    EXPECT_EQ(actions[t].item<int>(),
              results[params.capacity - params.n_steps + t].action);
    EXPECT_TRUE(std::abs(rewards[t].item<float>() -
                         results[params.capacity - params.n_steps + t].reward) <
                TEST_EPSILON);
    EXPECT_EQ(dones[t].item<bool>(),
              results[params.capacity - params.n_steps + t].done);
  }
}

TEST_P(TestDataBuffer, TestStoringAndRetrieval) {

  // Create the experiences at time t.
  auto experiences = getExperiences(observations, observations.size() - 1);

  // Create the multistep experiences at time t (experiences expected to be
  // returned by the replay buffer).
  auto results = getResultExperiences(observations, params.gamma,
                                      params.n_steps, 2 * params.capacity);

  // Fill the buffer with experiences.
  int n_experiences = params.capacity + params.n_steps - 1;
  for (int t = 0; t < n_experiences; t++) {
    buffer->append(experiences[t]);
  }

  // Check that experiences in the frame buffer are as expected.
  auto indices = torch::arange(params.capacity);
  auto [actions, rewards, dones] = (*buffer)[indices];
  for (int t = 0; t < params.capacity; t++) {
    EXPECT_EQ(actions[t].item<int>(), results[t].action);
    EXPECT_TRUE(std::abs(rewards[t].item<float>() - results[t].reward) <
                TEST_EPSILON);
    EXPECT_EQ(dones[t].item<bool>(), results[t].done);
  }

  // Keep pushing experiences to the buffer, effectively replacing all
  // experiences in the frame buffer.
  for (int t = 0; t < params.capacity; t++) {
    buffer->append(experiences[t + n_experiences]);
  }

  // Check that the new experiences in the frame buffer are as expected.
  std::tie(actions, rewards, dones) = (*buffer)[indices];
  for (int t = 0; t < params.capacity; t++) {
    EXPECT_EQ(actions[t].item<int>(), results[t + params.capacity].action);
    EXPECT_TRUE(std::abs(rewards[t].item<float>() -
                         results[t + params.capacity].reward) < TEST_EPSILON);
    EXPECT_EQ(dones[t].item<bool>(), results[t + params.capacity].done);
  }
}

TEST_P(TestDataBuffer, TestSaveAndLoad) {

  // Create the experiences at time t.
  auto experiences = getExperiences(observations, observations.size() - 1);

  // Fill the buffer with experiences.
  int n_experiences = params.capacity + params.n_steps - 1;
  for (int t = 0; t < n_experiences; t++) {
    buffer->append(experiences[t]);
  }

  // Save the data buffer.
  std::stringstream ss;
  buffer->save(ss);

  // Load the data buffer.
  auto loaded_buffer = DataBuffer(10, 10, 10, 10, 10);
  loaded_buffer.load(ss);

  // Check that the saved and loaded data buffers are identical.
  EXPECT_EQ(*buffer, loaded_buffer);
}

INSTANTIATE_TEST_SUITE_P(
    UnitTests, TestDataBuffer,
    testing::Values(
        DataBufferParameters(5, 1, 1), DataBufferParameters(5, 1, 0.9),
        DataBufferParameters(5, 2, 1), DataBufferParameters(5, 2, 0.99),
        DataBufferParameters(6, 2, 0.95), DataBufferParameters(7, 1, 0.5),
        DataBufferParameters(8, 3, 0.75), DataBufferParameters(9, 2, 0.8),
        DataBufferParameters(5, 6, 0.979), DataBufferParameters(5, 1, 0.98),
        DataBufferParameters(5, 1, 0.999), DataBufferParameters(9, 8, 0.1)));
} // namespace relab::test::agents::memory
